File: 231215/숨은 단어 찾기 2/find-hidden-words-2.js
const fs = require('fs');
const inputAll = () => fs.readFileSync('/dev/stdin').toString().trim().split('\n');
const print = (...args) => console.log(...args);

let [size, ...arr] = inputAll();

const [N, M] = size.split(' ').map(elm => Number(elm));
const CHAR_LIST = arr.map(line => Array.from(line));

function inRange(r, c){
    return 0 <= r && r < N && 0 <= c && c < M;
}

function zip(arr1, arr2){
    return Array.from({length: Math.min(arr1.length, arr2.length)}, (_, i) => [arr1[i], arr2[i]]);
}

function exploreLEE(CHAR_LIST, current, diff, length){
    const [cr, cc] = current;
    const [dr, dc] = diff;

    if(length === 3) return true;

    const [nr, nc] = [cr + dr, cc + dc];

    if(inRange(nr, nc) && CHAR_LIST[nr][nc] === 'E') return exploreLEE(CHAR_LIST, [nr, nc], diff, length + 1);

    return false;
}

function solution(CHAR_LIST){
    let count = 0;

    const drs = [0, -1, 0, 1, -1, -1, 1, 1];
    const dcs = [-1, 0, 1, 0, -1, 1, -1, 1];

    for(let i = 0; i < N; i++){
        for(let j = 0; j < M; j++){
            if(CHAR_LIST[i][j] === 'L'){
                for(const diff of zip(drs, dcs)){
                    if(exploreLEE(CHAR_LIST, [i, j], diff, 1)){
                        count++;
                    }
                }
            }
        }
    }

    return count;
}

print(solution(CHAR_LIST));

File: 231215/숨은 단어 찾기 2/find-hidden-words-2.py
# LEE의 개수를 구하면 된다.
# 전 8방향으로의 탐색이 필요
# drs, dcs를 활용해서 작성


def inRange(r, c):
    return 0 <= r < N and 0 <= c < M

def exploreLEE(CHAR_LIST, current, diff, length = 1):
    cr, cc = current
    dr, dc = diff

    nr, nc = cr + dr, cc + dc

    if length == 3:
        return True

    if not inRange(nr, nc):
        return False

    if CHAR_LIST[nr][nc] == 'E':
        return exploreLEE(CHAR_LIST, (nr, nc), diff, length + 1)

    return False

def solution(CHAR_LIST):
    count = 0

    drs = [0, -1, 0, 1, -1, -1, 1, 1]
    dcs = [-1, 0, 1, 0, -1, 1, -1, 1]

    for i in range(N):
        for j in range(M):
            # i, j를 시작으로 하는 것
            if CHAR_LIST[i][j] == 'L':
                for diff in zip(drs, dcs):
                   if exploreLEE(CHAR_LIST, (i, j), diff, 1):
                    count += 1

    return count

N, M = map(int, input().split())
CHAR_LIST = [list(input()) for _ in range(N)]

print(solution(CHAR_LIST))